### PR TITLE
installconfig: separate control plane and compute machine pools

### PR DESCRIPTION
**Pull request by staebler**
There should always be a control plane machine pool, so the declaration is moved from a generic `machines` field to a specific `controlPlane` field. The `machines` field contains only `compute` machines, so it is renamed to `compute`.

This is the first step in replacing references to master and worker with references to control plane and compute throughout the installer.

### CODE REVIEW

File: pkg/tfvars/tfvars.go [hunk 1]
```diff
@@ -108,3 +92,12 @@ func TFVars(clusterID string, cfg *types.InstallConfig, osImage, bootstrapIgn, m
 
 	return json.MarshalIndent(config, "", "  ")
 }
+
+func workerPool(pools []types.MachinePool) *types.MachinePool {
```
Comment (abhinavdahiya) [inline pkg/tfvars/tfvars.go]: we have 2 defintions of this function. is a common helper more appropiate? find name in compute list

File: pkg/asset/installconfig/installconfig.go [hunk 2]
```diff
@@ -55,7 +57,7 @@ func (a *InstallConfig) Generate(parents asset.Parents) error {
 
 	a.Config = &types.InstallConfig{
 		TypeMeta: metav1.TypeMeta{
-			APIVersion: "v1beta1",
+			APIVersion: "v1beta2",
```
Comment (wking) [inline pkg/asset/installconfig/installconfig.go]: I think we want a public constant for "the current version".  For an example, see [here][1] and other changes in that PR.

[1]: https://github.com/openshift/installer/pull/1154/files#diff-02200a3eac2ac223b4117e690cf2d0c7R23

**Comment by abhinavdahiya** [comment]
regarding https://github.com/openshift/installer/pull/1157/commits/fabb9119637ab71c7642ba9b44c2c57e5df73814 I think converters must exist in `pkg/types` as part of the api.

File: pkg/types/defaults/installconfig.go [hunk 3]
```diff
@@ -44,39 +43,27 @@ func SetInstallConfigDefaults(c *types.InstallConfig) {
...
-				Name:     "master",
-				Replicas: &numberOfMasters,
-			},
+	if c.ControlPlane == nil {
+		c.ControlPlane = &types.MachinePool{
+			Replicas: &defaultReplicaCount,
+		}
+	}
+	c.ControlPlane.Name = "master"
+	if c.Compute == nil {
```
Comment (wking) [inline pkg/types/defaults/installconfig.go]: Do we want to guard against no-workers until [the router issue][1] gets sorted?  If so, this should be `if len(c.Compute) == 0 {`.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1671136#c1

File: pkg/types/machinepools.go [hunk 4]
```diff
@@ -9,6 +9,8 @@ import (
 // MachinePool is a pool of machines to be installed.
 type MachinePool struct {
 	// Name is the name of the machine pool.
+	// For the control plane machine pool, the name will always be "master".
+	// For the compute machine pools, the only valid name is "worker".
```
Comment (wking) [inline pkg/types/machinepools.go]: It would be nice to lift this restriction.  All we're doing for compute nodes is pushing MachineSets which reference an installer-created worker IAM instance profile, right?  We can do that regardless of the name we set in the MachineSet.

File: pkg/asset/installconfig/installconfig.go [hunk 5]
```diff
@@ -142,3 +155,70 @@ func (a *InstallConfig) setDefaults() error {
 	defaults.SetInstallConfigDefaults(a.Config)
 	return nil
 }
+
+func loadDeprecated(data []byte, config *types.InstallConfig) error {
```
Comment (wking) [inline pkg/asset/installconfig/installconfig.go]: This is fairly compact, and we want to support it at some point, but it's not a strict requirement at the moment, with @crawford saying:

> We don't have to maintain backward compat for a beta release. We can just go to v1beta2 and tell people to update.

So if it helps review, we may want to spin this off into its own PR.

File: pkg/asset/machines/worker.go [hunk 6]
```diff
@@ -88,7 +88,7 @@ func (w *Worker) Generate(dependencies asset.Parents) error {
 	}
 
 	ic := installconfig.Config
-	pool := workerPool(ic.Machines)
+	pool := workerPool(ic.Compute)
```
Comment (wking) [inline pkg/asset/machines/worker.go]: This can be:

```go
for _, pool := range ic.Compute {
```

couldn't it?

File: pkg/tfvars/tfvars.go [hunk 7]
```diff
@@ -34,48 +34,32 @@ func TFVars(clusterID string, cfg *types.InstallConfig, osImage, bootstrapIgn, m
...
+			MasterRootVolume: aws.MasterRootVolume{
+				IOPS: cfg.ControlPlane.Platform.AWS.EC2RootVolume.IOPS,
+				Size: cfg.ControlPlane.Platform.AWS.EC2RootVolume.Size,
+				Type: cfg.ControlPlane.Platform.AWS.EC2RootVolume.Type,
+			},
+		}
+		if p := workerPool(cfg.Compute); p != nil {
+			if p.Platform.AWS != nil {
+				config.AWS.Worker = aws.Worker{
+					IAMRoleName: p.Platform.AWS.IAMRoleName,
```
Comment (wking) [inline pkg/tfvars/tfvars.go]: If #1154 lands first, we'll be able to drop this section.

File: pkg/tfvars/tfvars.go (hunk 1 above)
Comment (wking) [reply pkg/tfvars/tfvars.go]: > we have 2 defintions of this function. is a common helper more appropiate?

I think we should drop the function entirely, and just iterate through compute pools when generating MachineSets.

File: pkg/types/validation/installconfig.go [hunk 8]
```diff
@@ -103,21 +108,28 @@ func validateClusterNetwork(cn *netopv1.ClusterNetwork, fldPath *field.Path, ser
 	return allErrs
 }
 
-func validateMachinePools(pools []types.MachinePool, fldPath *field.Path, platform string) field.ErrorList {
+func validateControlPlane(pool *types.MachinePool, fldPath *field.Path, platform string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if pool.Name != "master" {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), pool.Name, []string{"master"}))
+	}
+	allErrs = append(allErrs, ValidateMachinePool(pool, fldPath, platform)...)
```
Comment (wking) [inline pkg/types/validation/installconfig.go]: We can extract the current "number of replicas must be positive" check from `ValidateMachinePool` and put it in this control-plane-specific validator instead.  And for workers, we have a check in `validateCompute` asserting that at least one `replicas` is strictly positive (the rest can be zero without causing issues).

File: pkg/types/validation/installconfig.go [hunk 9]
```diff
@@ -103,21 +108,28 @@ func validateClusterNetwork(cn *netopv1.ClusterNetwork, fldPath *field.Path, ser
...
+	return allErrs
+}
+
+func validateCompute(pools []types.MachinePool, fldPath *field.Path, platform string) field.ErrorList {
 	allErrs := field.ErrorList{}
 	poolNames := map[string]bool{}
 	for i, p := range pools {
```
Comment (wking) [inline pkg/types/validation/installconfig.go]: We should also check that `len(pools) > 0`, to work around [the router issue][1].

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1671136#c1

**Comment by wking** [comment]
I am super-excited about this :).  CC @dgoodwin, @vrutkovs about the pending install-config schema change.

**Comment by dgoodwin** [comment]
Awesome, and looks like it matches https://github.com/openshift/hive/pull/164 nicely. CC @abutcher

File: pkg/types/defaults/installconfig.go (hunk 3 above)
Comment (staebler) [reply pkg/types/defaults/installconfig.go]: Yes, missed this when rebasing with your changes. I will change it.

File: pkg/types/defaults/installconfig.go [hunk 10]
```diff
@@ -44,39 +43,27 @@ func SetInstallConfigDefaults(c *types.InstallConfig) {
...
-				}
-			default:
-				if c.Machines[i].Replicas == nil {
-					c.Machines[i].Replicas = pointer.Int64Ptr(0)
-				}
-			}
+	}
+	for i, p := range c.Compute {
+		if p.Replicas == nil {
+			c.Compute[i].Replicas = &defaultReplicaCount
```
[the remark this answers is not in the document]
Comment (staebler) [reply pkg/types/defaults/installconfig.go]: At the moment, the only compute machine pool accepted by the validator is "worker". My expectation in the future is that the installer will not care what the name of the compute machine pools are. The installer will not treat the "worker" or "compute" machine pool any different than any other compute machine pool. Do you have different expectations?

File: pkg/types/machinepools.go (hunk 4 above)
Comment (staebler) [reply pkg/types/machinepools.go]: Yes, this restriction will be lifted. This is just the first step in a larger transition that will include that lifted restriction.

File: pkg/asset/installconfig/installconfig.go (hunk 5 above)
Comment (staebler) [reply pkg/asset/installconfig/installconfig.go]: This is just in there temporarily so that this PR can land without breaking hive and byo in the near term. If it is spun off into its own PR, then that other PR would need to land before this PR.

File: pkg/typesv1beta1/doc.go [hunk 11]
```diff
@@ -0,0 +1,3 @@
+// Package typesv1beta1 defines v1beta1 structures for installer configuration and
```
[the remark this answers is not in the document]
Comment (staebler) [reply pkg/typesv1beta1/doc.go]: Yes, we need to structure `pkg/types` better to handle support for previous versions. I did not want to tackle that as part of this PR. Again, the support for v1beta1 types is only temporary so that this PR will not break people immediately. The `pkg/types/v1beta1' will be removed shortly after this PR lands.

File: pkg/asset/machines/worker.go (hunk 6 above)
Comment (staebler) [reply pkg/asset/machines/worker.go]: Ultimately, yes. But baby steps. And the validator only accepts a single compute machine pool named "worker" at the moment anyway.

File: pkg/types/validation/installconfig.go (hunk 8 above)
Comment (staebler) [reply pkg/types/validation/installconfig.go]: Sure. We can do that once we support more than a single machine pool named "worker".

File: pkg/types/validation/installconfig.go (hunk 9 above)
Comment (staebler) [reply pkg/types/validation/installconfig.go]: Yes. I will change that.

File: pkg/types/validation/installconfig.go (hunk 8 above)
Comment (staebler) [reply pkg/types/validation/installconfig.go]: Done.

**Comment by staebler** [comment]
> regarding [fabb911](https://github.com/openshift/installer/commit/fabb9119637ab71c7642ba9b44c2c57e5df73814) I think converters must exist in `pkg/types` as part of the api.

@abhinavdahiya Yes, I agree. I did not want to tackle doing long-term converters as part of this PR. I only wanted to put something in place temporarily so that current users such as hive and byo have a chance to switch over to v1beta2 without breaking. Once people have had a chance to make the switch, all the v1beta1 types and converted will be removed.

**Comment by wking** [comment]
@dgoodwin, @vrutkovs, do you need short-term compat with v1beta1, or can you bump your generators when you bump installer versions?

**Comment by dgoodwin** [comment]
I think we'll be ok, we'll adjust as soon as we vendor this in.

**Comment by vrutkovs** [comment]
@wking our GCP test was scrapped, scaleup test uses AWS cluster, so any option is fine by me

**Comment by staebler** [comment]
OK. If nobody needs the v1beta1 support, then I will remove it. Makes things easier.

File: pkg/tfvars/tfvars.go (hunk 1 above)
Comment (staebler) [reply pkg/tfvars/tfvars.go]: We will eventually drop this function very soon as part of subsequent steps of this change from master and worker to control plane and compute.

File: pkg/types/machinepools.go [hunk 12]
```diff
@@ -9,6 +9,8 @@ import (
 // MachinePool is a pool of machines to be installed.
 type MachinePool struct {
 	// Name is the name of the machine pool.
+	// For the control plane machine pool, the name will always be "master".
```
[the remark this answers is not in the document]
Comment (staebler) [reply pkg/types/machinepools.go]: As a data point, we had this conversation in hive back when I was on that team. We could not come up with a compelling use case for the user to provide a custom name for the control plane machines.

File: pkg/tfvars/tfvars.go (hunk 1 above)
Comment (staebler) [reply pkg/tfvars/tfvars.go]: The `workerPool` function has been dropped from `pkg/tfvars/tfvars.go` as it is no longer used.

**Comment by wking** [comment]
[e2e-aws][1]:

```
Failing tests:

[Feature:AnnotationTrigger] Annotation trigger reconciles after the image is overwritten [Suite:openshift/conformance/parallel]
[Feature:Builds] build with empty source  started build should build even with an empty source in build config [Suite:openshift/conformance/parallel]
[Feature:Builds] forcePull should affect pulling builder images  ForcePull test case execution docker [Suite:openshift/conformance/parallel]
[Feature:Builds][Conformance] build can reference a cluster service  with a build being created from new-build should be able to run a build that references a cluster service [Suite:openshift/conformance/parallel/minimal]
[Feature:Builds][Conformance][valueFrom] process valueFrom in build strategy environment variables  should successfully resolve valueFrom in s2i build environment variables [Suite:openshift/conformance/parallel/minimal]
[Feature:Builds][pruning] prune builds based on settings in the buildconfig  should prune errored builds based on the failedBuildsHistoryLimit setting [Suite:openshift/conformance/parallel]
[Feature:DeploymentConfig] deploymentconfigs generation [Conformance] should deploy based on a status version bump [Suite:openshift/conformance/parallel/minimal]
[Feature:DeploymentConfig] deploymentconfigs should respect image stream tag reference policy [Conformance] resolve the image pull spec [Suite:openshift/conformance/parallel/minimal]
[Feature:DeploymentConfig] deploymentconfigs with enhanced status [Conformance] should include various info in status [Suite:openshift/conformance/parallel/minimal]
[Feature:ImageLayers][registry] Image layer subresource should identify a deleted image as missing [Suite:openshift/conformance/parallel]
[k8s.io] Container Runtime blackbox test when starting a container that exits should report termination message if TerminationMessagePath is set [NodeConformance] [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-api-machinery] Generated clientset should create v1beta1 cronJobs, delete cronJobs, watch cronJobs [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-apps] DisruptionController evictions: enough pods, absolute => should allow an eviction [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-cli] Kubectl client [k8s.io] Kubectl describe should check if kubectl describe prints relevant information for rc and pods  [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
[sig-network] NetworkPolicy NetworkPolicy between server and client should enforce multiple, stacked policies with overlapping podSelectors [Feature:NetworkPolicy] [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Pre-provisioned PV (ext4)] volumes should be mountable [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: hostPathSymlink] [Testpattern: Inline-volume (default fs)] subPath should support existing single file [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: hostPath] [Testpattern: Inline-volume (default fs)] subPath should support non-existent path [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: hostPath] [Testpattern: Pre-provisioned PV (default fs)] subPath should support readOnly directory specified in the volumeMount [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] In-tree Volumes [Driver: vSphere] [Testpattern: Dynamic PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] PVC Protection Verify that scheduling of a pod that uses PVC that is being deleted fails and the pod becomes Unschedulable [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-storage] Projected should provide container's cpu limit [NodeConformance] [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
[sig-storage] Projected should provide podname as non-root with fsgroup and defaultMode [NodeFeature:FSGroup] [Suite:openshift/conformance/parallel] [Suite:k8s]
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1157/pull-ci-openshift-installer-master-e2e-aws/3398/build-log.txt

**Comment by staebler** [comment]
@wking @abhinavdahiya CI is green. Do either of you have any further review comments that you would like to add?

File: pkg/asset/ignition/machine/master_test.go [hunk 13]
```diff
@@ -30,11 +31,9 @@ func TestMasterGenerate(t *testing.T) {
 					Region: "us-east",
 				},
 			},
-			Machines: []types.MachinePool{
-				{
-					Name:     "master",
-					Replicas: func(x int64) *int64 { return &x }(3),
-				},
+			ControlPlane: &types.MachinePool{
+				Name:     "master",
```
Comment (derekwaynecarr) [inline pkg/asset/ignition/machine/master_test.go]: is there a desire to eliminate the term "master" generally?

Comment (derekwaynecarr) [reply pkg/asset/ignition/machine/master_test.go]: just asking as that is a broader change to push through.

Comment (staebler) [reply pkg/asset/ignition/machine/master_test.go]: Yes. The broader change of replacing the terms master and worker with control-plane and compute are tackled in #1227.

**Comment by flaper87** [comment]
If you don't mind, I'd like to test this patch with OpenStack before it goes in. I'll wait for the next rebase to test it.

**Comment by staebler** [comment]
> If you don't mind, I'd like to test this patch with OpenStack before it goes in. I'll wait for the next rebase to test it.

A test on OpenStack sounds wonderful. The PR has been rebased.

File: pkg/types/installconfig.go [hunk 14]
```diff
@@ -47,11 +47,14 @@ type InstallConfig struct {
...
 	*Networking `json:"networking,omitempty"`
 
-	// Machines is the list of MachinePools that need to be installed.
+	// ControlPlane is the configuration for the machines that comprise the
+	// control plane.
 	// +optional
-	// Default on AWS and OpenStack is 3 masters and 3 workers.
-	// Default on Libvirt is 1 master and 1 worker.
-	Machines []MachinePool `json:"machines,omitempty"`
+	ControlPlane *MachinePool `json:"controlPlane,omitempty"`
```
Comment (wking) [inline pkg/types/installconfig.go]: We'll need to bump [`InstallConfigVersion`][1] for this.  Possibly also with a temporary hack to get through CI (e.g. like the one being removed by #1251).

[1]: https://github.com/openshift/installer/blob/e1082a3ba20036be273300984ac01b891993764b/pkg/types/installconfig.go#L14

Comment (staebler) [reply pkg/types/installconfig.go]: I wasn't sure whether the version needed to be bumped with every change. Or whether the version change covered the collections of API changes between beta1 and beta2.

Comment (wking) [reply pkg/types/installconfig.go]: We cut 0.12.0 with v1beta2, so I think we need another bump.  But I'm fine combining future changes that land between releases.

**Comment by staebler** [comment]
/hold cancel

The feature exception for this has been approved.

File: pkg/types/installconfig.go (hunk 14 above)
Comment (staebler) [reply pkg/types/installconfig.go]: Fair enough. I will bump up to v1beta3.

Comment (abhinavdahiya) [inline pkg/types/installconfig.go]: Why is this omitempty? and nilable.

from the commit message **`There should always be a control plane machine pool`**

Comment (wking) [reply pkg/types/installconfig.go]: Because the installer can fill in the default.

Comment (staebler) [reply pkg/types/installconfig.go]: It is omit empty and nilable because the user does not have to provide it. If it is omitted, then the installer will add default values to the InstallConfig used. The validator will ensure that the field is not nil in the final InstallConfig.

Comment (abhinavdahiya) [reply pkg/types/installconfig.go]: :confused: sure
